test/clj_puppetdb/http_test.clj

```diff
@@ -52,23 +52,57 @@
         response-data ["node-1" "node-2"]
         response-data-encoded (json/encode response-data)
         response-headers {"x-records" (.toString (count response-data))}
-        fake-get (fn[status] {:status status :body (io/input-stream (.getBytes response-data-encoded)) :headers response-headers})]
+        fake-get (fn [status] {:status status :body (io/input-stream (.getBytes response-data-encoded)) :headers response-headers})]
 
-      (testing "Should have proper response"
-        (with-redefs [http/request-with-client (fn[_ _ _] (future (fake-get 200)))]
-          (let [GET-response (GET client path params)]
-            (is (= (first GET-response) response-data))
-            (is (= (second GET-response) response-headers)))))
+    (testing "Should have proper response"
+      (with-redefs [http/request-with-client (fn [_ _ _] (future (fake-get 200)))]
+        (let [GET-response (GET client path params)]
+          (is (= (first GET-response) response-data))
+          (is (= (second GET-response) response-headers)))))
 
-      (testing "Should throw proper exception"
-        (with-redefs [http/request-with-client (fn[_ _ _] (future (fake-get 400)))]
-          (try
-            (GET client path params)
-            (catch ExceptionInfo ei
-              (let [info (.getData ei)]
-                (is (= (:status info) 400))
-                (is (= (:kind info) :puppetdb-query-error))
-                (is (= (:params info) params))
-                (is (= (:endpoint info) path))
-                (is (= (:host info) host))
-                (is (= (:msg info) response-data-encoded)))))))))
+    (testing "Should throw proper exception"
+      (with-redefs [http/request-with-client (fn [_ _ _] (future (fake-get 400)))]
+        (try
+          (GET client path params)
+          (catch ExceptionInfo ei
+            (let [info (.getData ei)]
+              (is (= (:status info) 400))
+              (is (= (:kind info) :puppetdb-query-error))
+              (is (= (:params info) params))
+              (is (= (:endpoint info) path))
+              (is (= (:host info) host))
+              (is (= (:msg info) response-data-encoded)))))))))
+
+(deftest catching-exceptions-test
+  (testing "Should pass"
+    (is (= (catching-exceptions ((constantly {:body "foobar"})) (assoc-kind {} :something-bad-happened)) {:body "foobar"})))
+  (testing "Should rethrow proper exception on an exception"
+    (try
+      (catching-exceptions (#(throw (NullPointerException.))) (assoc-kind {} :something-bad-happened))
+      (is (not "Should never get to this place!!"))
+      (catch ExceptionInfo ei
+        (let [info (.getData ei)]
+          (is (= (:kind info) :something-bad-happened))
+          (is (instance? NullPointerException (:exception info)))))))
```

Comment:
For context: @scottyw wandered over to my desk and explained to me that it's not that it's a dependency in the project.clj that was an issue, but rather that by using slingshot, we make anything that uses library basically require slingshot for doing it's error handling. While I can understand that argument, it seems like we're making life maybe slightly easier for potential, mostly theoretical users, at the expense of making our lives slightly harder. This seems especially true given that you can catch slingshot errors using a normal `try` without slingshot:

~~~clojure
$ lein repl
user=> (require '[slingshot.slingshot :refer [throw+]])
nil
user=> (try (throw+ {:error "foo"}) (catch Exception e (println e)))
#<ExceptionInfo clojure.lang.ExceptionInfo: throw+: {:error "foo"} {:error foo}>
nil
~~~

So I'm definitely :+1:  on those tech-debt tickets.